racket/collects/racket/private/port.rkt

```diff
@@ -110,37 +110,36 @@
       [(mode) (file-stream-buffer-mode p mode)]
       [() (file-stream-buffer-mode p)]))))
 
-(define (copy-port src dest . dests)
+(define (copy-port src dest . dests*)
   (unless (input-port? src)
     (raise-type-error 'copy-port "input-port" src))
   (for-each
    (lambda (dest)
      (unless (output-port? dest)
        (raise-type-error 'copy-port "output-port" dest)))
-   (cons dest dests))
-  (let ([s (make-bytes 4096)]
-        [dests (cons dest dests)])
-    (let loop ()
-      (let ([c (read-bytes-avail! s src)])
-        (cond
-          [(number? c)
-           (let loop ([dests dests])
-             (unless (null? dests)
-               (let loop ([start 0])
-                 (unless (= start c)
-                   (let ([c2 (write-bytes-avail s (car dests) start c)])
-                     (loop (+ start c2)))))
-               (loop (cdr dests))))
-           (loop)]
-          [(procedure? c)
-           (let ([v (let-values ([(l col p) (port-next-location src)])
-                      (c (object-name src) l col p))])
-             (let loop ([dests dests])
-               (unless (null? dests)
-                 (write-special v (car dests))
-                 (loop (cdr dests)))))
-           (loop)]
-          [else
-           ;; Must be EOF
-           (void)])))))
+   (cons dest dests*))
+
+  (define sz 4096)
+  (define s (make-bytes sz))
+  (define dests (cons dest dests*))
+
+  (let loop ()
+    (define c (read-bytes-avail! s src))
+    (cond
+      [(number? c)
+       (for ([dest (in-list dests)])
+         (let write-loop ([bytes-written 0])
+           (unless (= bytes-written c)
+             (define c2 (write-bytes-avail s dest bytes-written c))
+             (write-loop (+ bytes-written c2)))))
+       (loop)]
+    [(procedure? c)
+     (let ([v (let-values ([(l col p) (port-next-location src)])
```

Comment:
This could be rewritten as 

```racket
(define-values (l col p) (port-next-location src))
(define v (c (object-name src) l col p))
(for ...)
```

Reply:
True, I like that. I will push a new patch.